File: Knou/2023년 2학기/simulation/java_src(file)/EX1_5.java
/**
 * 주사위 시뮬레이션
 * seed   : 난수의 초기값으로 9965
 * COUNT  : 각 숫자들의 출현횟수를 계산하기 위한 배열
 * side   : 주사위의 숫자, 즉 1..6
 * U      : [0,1)범위의 일양분포의 난수
 * nrolls : 시행횟수
 */
import java.lang.*;
import java.io.*;

class Rolling {
       final static int nrolls=3000;
       final static int SEED=9965;
       public float up;
       public int count[] = new int[6];

       public Rolling() {
               int i;
               for(i=0; i<5; ++i) count[i]=0;
       }

       public int random(int np, float u){
               np = np * 843314861 + 453816693;
               if(np < 0){
                       np = np + 2147483647;
                       np = np + 1;
               }
               up = (float)(np * 0.4656612e-9);
               return np;
       }

       public void ComputeRolling(){
               int i, side, n=SEED;
               PrintWriter out=null;

               i = 0;
               while (i < nrolls) {
                      n = random(n, up);
                      side = (int) Math.floor(up * 6);
                      count[side] = count[side] + 1;
                      i++;
               }

               try {
                   File f = new File("C:\\WORK\\EX1_5.OUT");
                   FileWriter fw = new FileWriter(f);
                   BufferedWriter bw = new BufferedWriter(fw);
                   out = new PrintWriter(bw);

                   out.println("==========================");
                   out.println("      THE STATISTICS");
                   out.println("==========================");
                   for (i = 0; i < 6; ++i)
                        out.println(" SIDE " + (i + 1) + " OCCURED " + count[i] + " TIMES");
               } catch(IOException ioe) { }
                 finally {
                       if (out != null) out.close();
                 }
       }
}

public class EX1_5 {
       public static void main(String[] args) {
                   Rolling r = new Rolling();
                   r.ComputeRolling();
  }
}

File: Knou/2023년 2학기/시뮬레이션/VC++2019/ex5/ex1-5.c
// ROLLING A DIE SIMULATION                           //

// NROLLS : total number of rolls                     //
// COUNT  : occurrences of each side                  //
// SEED   : seed for random number generator          //
// U      : uniform random number [0,1)               //

// PROGRAM  EX1-5                                     //

#include<stdio.h>
#include<math.h>

#define nrolls   3000
#define SEED    35213


//  RANDOM NUMBER GENERATOR //
void random(long *pn,float *pu)
{
	*pn = *pn * 843314861 + 453816693;
	if(*pn < 0)
	       { *pn = *pn +2147483647;
		 *pn = *pn + 1;
	       }
	*pu = *pn * 0.4656612e-9;
	return;
}


//  MAIN  PROGRAM  //
void main()
{
       int count[6], side, i;
       long n=SEED;
       float u=0;
       FILE *output;
       output = fopen("c:\\work\\ex1-5.out","wt");

       for (i = 0; i < 6; ++i)   count[i]=0;
       i=0;
	    while (i < nrolls)
	    {
		random(&n,&u);
		side = floor(u * 6) ;
		count[side] = count[side] + 1;
		++i;
	    }
       fprintf(output,"         THE STATISTICS\n\n");
	    for (i = 0 ; i<6 ;++i)
		fprintf(output,"SIDE   %d  OCCURED    %d    TIMES\n",i+1,count[i]);
       fclose(output);
}
